### PR TITLE
[SPARK-20056][ML] IsotonicRegression support Numeric features

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
IsotonicRegression support Numeric features

## How was this patch tested?
existing tests and added tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/regression/IsotonicRegression.scala [hunk 1]
```diff
@@ -84,7 +84,7 @@ private[regression] trait IsotonicRegressionBase extends Params with HasFeatures
       val extract = udf { v: Vector => v(idx) }
       extract(col($(featuresCol)))
     } else {
-      col($(featuresCol))
+      col($(featuresCol)).cast(DoubleType)
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/regression/IsotonicRegression.scala]: would there be precision loss by doing this?

File: mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala [hunk 2]
```diff
@@ -189,6 +191,47 @@ class IsotonicRegressionSuite
...
+      "Column features must be of type NumericType or VectorUDT," +
+        " but was actually of type StringType"))
+
+    val thrown2 = intercept[IllegalArgumentException] {
+      expected.transform(dfWithStringFeatures)
+    }
+    assert(thrown2.getMessage.contains(
+      "Column features must be of type NumericType or VectorUDT," +
+        " but was actually of type StringType"))
+  }
```
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala]: Why not use [MLTestingUtil's checkNumericTypes](https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala#L41-L88)?

**Review by BenFradet — COMMENTED** [review]
I have a small Q regarding the tests

File: mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala (hunk 2 above)
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala]: `checkNumericTypes` only test `fit` method on `labelCol` and `weightCol`.
`IsotonicRegressor` is the only algorithm which need to test both `fit` and `transform` on `featureCol`. So I prefer not to use `checkNumericTypes`. Thanks for your review

File: mllib/src/main/scala/org/apache/spark/ml/regression/IsotonicRegression.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/regression/IsotonicRegression.scala]: Precision loss may happen here. But I think we can not avoid it now, for we use `mllib.IsotonicRegression` and the input dataframe is of type `input: RDD[(Double, Double, Double)])`

**Comment by MLnick** [comment]
For the other algos we've done something similar - but aimed to allowing any numeric type for `label` and `weight` columns. I'm not totally against this here, but I think it falls into a slightly different "category" - i.e. supporting different numeric types for input features.

Overall I think it's more natural to enforce a double feature column (if you think about it, a feature vector column is implicitly enforced to be double as we don't support arbitrary numeric types for vectors).

I'm not convinced this solves any major pain point.

File: mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala (hunk 2 above)
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala]: My bad, skimmed too quickly over the tests and it looked very similar to what I had done in the past.

**Comment by zhengruifeng** [comment]
@MLnick Agree. I will close this PR.

**Pull request closed without merging**